docs/src/jekyll/docs.md

```diff
@@ -0,0 +1,546 @@
+val cache = InMemoryCache(UserSource.identity(2) -> User(2, "@two"))
+val fut: Future[List[User]] = Fetch.run(fch, cache)
+
+await(fut)
+// ~~~>[thread: 180] Fetching users List(1, 3)
+// <~~~[thread: 180] Fetching users List(1, 3)
+
+//=> List(User(1,@egg_1), User(2,@two), User(3,@egg_3))
+```
+
+## Replaying a fetch
+
+When running a fetch we are generally interested in its final result. However, we also have access to the cache
+and information about the executed rounds once we run a fetch. Fetch's interpreter keeps its state in an environment
+(implementing the `Env` trait) and we can get both the environment and result after running a fetch using `Fetch.runFetch`
+instead of `Fetch.run`.
+
+Knowing this, we can replay a fetch reusing the cache of a previous one. The replayed fetch won't have to call any of the
+data sources.
+
+```scala
+val fch: Fetch[List[User]] = Fetch.traverse(List(1, 2, 3))(getUser)
+val env = await(Fetch.runEnv[Future](fch))
+// ~~~>[thread: 261] Fetching users List(1, 2, 3)
+// <~~~[thread: 261] Fetching users List(1, 2, 3)
+
+//=> List(User(1,@egg_1), User(2,@egg_2), User(3,@egg_3))
+
+val fut: Future[List[User]] = Fetch.run(fch, env.cache)
+
+await(fut)
+//=> List(User(1,@egg_1), User(2,@egg_2), User(3,@egg_3))
+```
+
+## Implementing a custom cache
+
+The default cache is implemented as an in-memory map, but users are free to use their own caches when running a fetch. Your cache should implement the `DataSourceCache` trait, and after that you can pass it to Fetch's `run` methods.
+
+```scala
+trait DataSourceCache {
+  def update[I, A](k: DataSourceIdentity, v: A): DataSourceCache
+  def get[I](k: DataSourceIdentity): Option[Any]
+}
+```
+
+Let's reimplement the in memory cache found in Fetch, we'll write a case class that'll store the cache contents in an in-memory immutable map and implement `DataSourceCache`.
+
+```scala
+case class MyInMemoryCache(state: Map[DataSourceIdentity, Any]) extends DataSourceCache {
+  override def get[I](k: DataSourceIdentity): Option[Any] =
+    state.get(k)
+
+  override def update[I, A](k: DataSourceIdentity, v: A): MyInMemoryCache =
+    copy(state = state.updated(k, v))
+}
+```
+
+Now that we have a data type, we can use it when running a fetch:
+
+```scala
+val cache = MyInMemoryCache(Map(UserSource.identity(1) -> User(1, "dialelo")))
+val fch: Fetch[User] = getUser(1)
+val fut: Future[User] = Fetch.run(fch, cache)
+
+await(fut)
+//=> User(1,dialelo)
+```
+
+# Error handling
+
+You probably have noticed that `DataSource.fetch` takes a list of identities and returns a map of identities to their result, taking
+into account the possibility of some identities not being found. Whenever an identity isn't found, the fetch execution will
+fail.
+
+## Diagnosing fetch failures
+
+Whenever a fetch fails, a `FetchFailure` exception is thrown. The `FetchFailure` will have the environment, which gives you information about the execution of a fetch.
```

Comment:
I'm guessing the exception is not actually thrown but captured by the MonadError instance which depending on your target may throw it like in the case of `Id` but other better fit ones like `Task` would place it on the Left.